arlo_server/audit_boards.py

```diff
@@ -187,3 +190,101 @@ def list_audit_boards(
+            rc for rc in round.round_contests if rc.contest_id == contest.id
+        )
+        round_contest.end_p_value = max(risk.values())
+        round_contest.is_complete = is_complete
+
+
+def end_round(election: Election, round: Round):
+    calculate_risk_measurements(election, round)
+    round.ended_at = datetime.utcnow()
+
+
+def is_round_complete(election: Election, round: Round) -> bool:
+    num_jurisdictions_without_audit_boards_set_up = (
+        # For each jurisdiction...
+        Jurisdiction.query.filter_by(election_id=election.id)
+        # Where there are ballots that haven't been audited...
+        .join(Jurisdiction.batches)
+        .join(Batch.ballots)
+        .filter(SampledBallot.vote.is_(None))
+        # And those ballots got sampled this round...
+        .join(SampledBallot.draws)
+        .filter_by(round_id=round.id)
+        # Count the number of audit boards set up.
+        .outerjoin(
+            AuditBoard,
+            and_(
+                AuditBoard.jurisdiction_id == Jurisdiction.id,
+                AuditBoard.round_id == round.id,
+            ),
+        )
+        .group_by(Jurisdiction.id)
+        # Finally, count how many jurisdictions have no audit boards set up.
+        .having(func.count(AuditBoard.id) == 0)
+        .count()
+    )
+    all_audit_boards_set_up = num_jurisdictions_without_audit_boards_set_up == 0
+    all_audit_boards_signed_off = all(
+        ab.signed_off_at is not None for ab in round.audit_boards
+    )
+    return all_audit_boards_set_up and all_audit_boards_signed_off
+
+
+SIGN_OFF_AUDIT_BOARD_REQUEST_SCHEMA = {
+    "type": "object",
+    "properties": {
+        "memberName1": {"type": "string"},
+        "memberName2": {"type": "string"},
+    },
+    "additionalProperties": False,
+    "required": ["memberName1", "memberName2"],
+}
+
+# Raises if invalid
+def validate_sign_off(sign_off_request: JSONDict, audit_board: AuditBoard):
+    validate(sign_off_request, SIGN_OFF_AUDIT_BOARD_REQUEST_SCHEMA)
+
+    for name in [sign_off_request["memberName1"], sign_off_request["memberName2"]]:
+        if name not in {audit_board.member_1, audit_board.member_2}:
+            raise BadRequest(f"Audit board member name did not match: {name}")
+
+    if any(b.vote is None for b in audit_board.sampled_ballots):
+        raise Conflict(f"Audit board is not finished auditing all assigned ballots")
+
+
+@app.route(
+    "/election/<election_id>/jurisdiction/<jurisdiction_id>/round/<round_id>/audit-board/<audit_board_id>/sign-off",
+    methods=["POST"],
+)
+@with_audit_board_access
+def sign_off_audit_board(
+    election: Election,
+    jurisdiction: Jurisdiction,  # pylint: disable=unused-argument
+    round: Round,
+    audit_board: AuditBoard,
+):
+    validate_sign_off(request.get_json(), audit_board)
+
+    audit_board.signed_off_at = datetime.utcnow()
+
+    if is_round_complete(election, round):
+        end_round(election, round)
+
+    db.session.commit()
```

Comment:
I wasn't totally sure whether the helper functions (`end_round`, `calculate_risk_measurement`) should each call `db.session.commit()` or whether it should only be called here once at the end. I opted for just calling it once because I figured if we wanted to rollback the changes, they should all be in one transaction. But not sure if that's actually what's going on under the hood or if that's the right approach.